compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp

```diff
@@ -260,6 +265,218 @@ static CodeGenPipeline getTensorCorePipeline(Type elementType) {
+        return failure();
+      foundSingleReductionOutput = true;
+      continue;
+    }
+    if (!op.getMatchingIndexingMap(&initOpOperand).isIdentity())
+      return failure();
+  }
+  if (!foundSingleReductionOutput)
+    return failure();
+
+  SmallVector<int64_t> workgroupTileSizes(op.getNumLoops(), 0);
+  SmallVector<int64_t> threadTileSizes(op.getNumLoops(), 0);
+  SmallVector<int64_t> partialReductionTileSizes(op.getNumLoops(), 0);
+  SmallVector<int64_t> threadCounts(op.getNumLoops(), 1);
+  SmallVector<int64_t> subGroupCounts(op.getNumLoops(), 1);
+  SmallVector<int64_t> mapping(op.getNumLoops());
+  std::iota(mapping.begin(), mapping.end(), 0);
+
+  int64_t lastReductionDim = reductionDims.back();
+
+  for (int64_t dim : parallelDims) {
+    workgroupTileSizes[dim] = 1;
+  }
+
+  for (int64_t dim : reductionDims) {
+    threadTileSizes[dim] = 1;
+    partialReductionTileSizes[dim] = 1;
+  }
+
+  int64_t reductionSize = 1;
+  for (int64_t dim : reductionDims)
+    reductionSize *= bounds[dim];
+
+  int64_t subgroupSize = 0;
+  for (int s : wgp.getSubgroupSizeChoices().asArrayRef()) {
+    if (reductionSize % s == 0) {
+      subgroupSize = s;
+      break;
+    }
+  }
+  if (subgroupSize == 0)
+    return failure();
+
+  auto bitWidth = getBitWidth(op);
+  if (failed(bitWidth)) {
+    return failure();
+  }
+
+  const std::optional<int64_t> maxLoadBits = wgp.getMaxLoadInstructionBits();
+  const unsigned largestLoadSizeInBits =
+      maxLoadBits.has_value() ? *maxLoadBits : 128;
+
+  int64_t threadLoads = largestLoadSizeInBits / *bitWidth;
+  if (reductionSize % threadLoads != 0)
+    return failure();
+
+  int64_t lastDimSize = bounds[reductionDims.back()];
+  int64_t partialReductionSize = subgroupSize * threadLoads;
+
+  while (lastDimSize % partialReductionSize != 0)
+    partialReductionSize >>= 1;
+
+  threadLoads = std::min(lastDimSize, threadLoads);
+
+  int64_t threadBasis = partialReductionSize / threadLoads;
+
+  partialReductionTileSizes[lastReductionDim] = partialReductionSize;
+  threadTileSizes[lastReductionDim] = threadLoads;
+  threadCounts[lastReductionDim] = threadBasis;
+
+  auto context = op.getContext();
+  Builder b(context);
+  SmallVector<NamedAttribute, 1> attrs;
+
+  ArrayAttr subgroupBasisAttr = b.getArrayAttr(
+      {b.getI64ArrayAttr(subGroupCounts), b.getI64ArrayAttr(mapping)});
+
+  ArrayAttr threadBasisAttr = b.getArrayAttr(
+      {b.getI64ArrayAttr(threadCounts), b.getI64ArrayAttr(mapping)});
+
+  attrs.emplace_back(b.getStringAttr("workgroup"),
+                     b.getI64ArrayAttr(workgroupTileSizes));
+  attrs.emplace_back(b.getStringAttr("partial_reduction"),
+                     b.getI64ArrayAttr(partialReductionTileSizes));
+  attrs.emplace_back(b.getStringAttr("thread"),
+                     b.getI64ArrayAttr(threadTileSizes));
+  attrs.emplace_back(b.getNamedAttr("thread_basis", threadBasisAttr));
+  attrs.emplace_back(b.getNamedAttr("subgroup_basis", subgroupBasisAttr));
```

Comment:
I don't get why you use small size of 1 for `attrs` and then add 5 elements immediately after. Either set a size that's the same or larger than the number of elements, or use 0 for no inline storage.

Also, to construct `NamedAttr`, you can provide the string as `StringRef` instead of `StringAttr`. See how we build it here: https://github.com/iree-org/iree/blob/ddce57784f8193f15d9275515bb90bbc7bc25a56/compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp#L428-L430